constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,479 @@
+
+
+# Returns the f_syscall operation flag which is set when SYSCALL control operation is executed.
+fn f_syscall(b: vector[7], extra: scalar) -> scalar:
+    return extra & !b[4] & b[3] & !b[2]
+
+
+# Returns the f_end operation flag which is set when END operation is executed.
+fn f_end(b: vector[7], extra: scalar) -> scalar:
+    return extra & b[4] & !b[3] & !b[2]
+
+
+# Returns the f_halt operation flag which is set when HALT operation is executed.
+fn f_halt(b: vector[7], extra: scalar) -> scalar:
+    return extra & b[4] & b[3] & b[2]
+
+
+# Returns the f_push operation flag which is set when PUSH operation is executed.
+fn f_push(b: vector[7], extra: scalar) -> scalar:
+    return extra & !b[4] & !b[3] & b[2]
+
+
+# Returns the f_ctrl flag which is set when any one of the control flow operations that has flags 
+# above (JOIN, SPLIT, LOOP, CALL, SYSCALL) is being executed.
+fn f_ctrl(b: vector[7], extra: scalar) -> scalar:
+    # flag for SPAN, JOIN, SPLIT, LOOP
+    let f_sjsl = b[6] & !b[5] & b[4] & b[3]
+
+    # flag for END, REPEAT, RESPAN, HALT
+    let f_errh = b[6] & b[5] & b[4]
+
+    return f_sjsl + f_errh + f_call(b, extra) + f_syscall(b, extra)
+
+
+# Returns f_ctrli flag which is set to 1 when a control flow operation that signifies the 
+# initialization of a control block is being executed on the VM.
+fn f_ctrli(b: vector[7], extra: scalar) -< scalar:
+    return f_join(b) + f_split(b) + f_loop(b) + f_call(b, extra) + f_syscall(b, extra)
+
+
+# Returns transition label, composed of the operation label and the periodic columns that uniquely 
+# identify each transition function.
+fn get_transition_label(op_label: scalar) -> scalar:
+    return op_label + 2^4 * cycle_row_7 + 2^5 * cycle_row_0
+
+
+# Returns f_g8 flag which is set to 1 if there are 8 operation groups in the batch.
+fn get_f_g8(op_batch_flags: vector[3]) -> scalar:
+    return op_batch_flags[0]
+
+
+# Returns f_g4 flag which is set to 1 if there are 4 operation groups in the batch.
+fn get_f_g4(op_batch_flags: vector[3]) -> scalar:
+    return !op_batch_flags[0] & op_batch_flags[1] & op_batch_flags[2]
+
+
+# Returns f_g2 flag which is set to 1 if there are 2 operation groups in the batch.
+fn get_f_g2(op_batch_flags: vector[3]) -> scalar:
+    return !op_batch_flags[0] & !op_batch_flags[1] & op_batch_flags[2]
+
+
+# Returns f_g1 flag which is set to 1 if there are 1 operation groups in the batch.
+fn get_f_g1(op_batch_flags: vector[3]) -> scalar:
+    return !op_batch_flags[0] & op_batch_flags[1] & !op_batch_flags[2]
+
+
+### Helper evaluators #############################################################################
+
+# Enforces that column must be binary.
+ev is_binary(main: [a]):
+    enf a^2 = a
+
+
+# Enforces that value in column is copied over to the next row.
+ev is_unchanged(main: [column]):
+    ev column' = column
+
+
+# Enforces decoder general constraints.
+ev general_constraints(main: [addr, op_bits[7], hasher[8], is_span, s0], aux: [extra]):
```

Comment:
I believe the `extra` column for degree reduction is the 23rd column at the end of the main trace. It's [documented in the code](https://github.com/0xPolygonMiden/miden-vm/blob/b3a1acb14603d58979026aeb9569c59623613cbc/processor/src/decoder/mod.rs#L299), but it's not shown in the execution trace diagrams in the decoder docs, which only show 22 columns.

This also suggests that there should be a constraint somewhere to enforce that `extra == 1` when b[6] & b[7]. I'm not sure if this constraint is described anywhere currently. Have you seen it?

@bobbinth anything to add or correct here?

Reply:
Yes, `extra` column is for degree reduction, which is named as `be` in the code. 

I didn't see constraints for this column in docs, but they seem clear, I'll add them in next commit.

Reply:
@Overcastan I think you misunderstood my comment. What I'm saying is that you have the `extra` column in the auxiliary trace, which is not correct. It belongs in the main trace, where it is the 23rd column. This is documented in the code, but it is not shown in the diagrams in the decoder documentation.

The constraint you've added looks correct though

Reply:
Can you please submit a PR to update the decoder constraint docs with the new constraint for the `extra` column as well?